CQL3 support

ctavan opened the pull request:
Cassandra 1.1 will have support for CQL 3 which is not backwards compatible to CQL 2, default of Cassandra 1.0.

CQL 2 will remain the default version in Cassandra 1.1 but users may choose the CQL version they want to use for a connection by issuing a `set_cql_version()` thrift-call.

This patch adds an optional config setting that lets the user specify the CQL version.

I had to recompile the thrift-generated JS libraries. Unfortunately thrift produces buggy files, see https://issues.apache.org/jira/browse/THRIFT-1297 That's why I have added two patches that correct for the thrift flaws. Of course these issues should rather be fixed in thrift than hacked here...

My changes are backwards-compatible with Cassandra 1.0 (tested with 1.0.8) as far as the test-suite is concerned, i.e. I'm not aware of any breaking changes.

I'll continue to contribute some CQL3 specific test cases during the next days.

Any comments are appreciated!

ctavan committed:
Add new 1.1.0-beta2 cassandra.thrift and recompile it

ctavan committed:
Add patches for broken thrift output

ctavan committed:
Use thrift api from C* 1.1

ctavan committed:
Add clean_compile make target to .PHONY list

ctavan committed:
Allow setting the cql version for a connection

ctavan committed:
Add note about cqlVersion in README

ctavan committed:
Move old thrift libs

ctavan committed:
Remove 1.0 thrift stuff and move 1.1 thrift stuff up one level

devdazed commented:
I've already made several changes to the thrift generated files in previous commits.  We need to make sure that all of those changes get moved to the new thrift generated files as well.  I've also heard rumors that C\* 1.1 will be using a different version of thrift.  If this is true we need to create the correct version of helenus-thrift.

also, is C\* 1.1 Thrift api backward compatible with C\* 1.0?  if not, we may want to do this in a branch to ensure compatibility.

I really like the idea of using .patch files to patch the thrift generated code.

ctavan commented:
As far as the Thrift API is concerned: The cassandra devs promised it will never be broken again (at least that's what they said at the cassandraEU conference).

At the moment they are still using thrift 0.7 in the cassandra-1.1 branch. I don't think that will change now since the 1.1 beta is already out. So we'll most probably be able to stick to thrift 0.7 (and I should maybe generate the JS files again using thrift 0.7 instead of thrift 0.8...)

Will you check if I did all the changes you had made to the thrift-generated output correctly or do you want me to do this? Could you point me to the relevant commits?

ctavan commented:
Concerning `this.count !== null || this.count !== undefined`: I blindly copied that from the old thrift-generated files but that's of course wrong. Gonna fix it in a second.

ctavan committed:
Fix operators when checking for null/undefined

Original commit is aaca8447fcba7e07c44891a904a559fdf375a94d

ctavan commented:
OK, I've had a look at your changes to the autogenerated files:
- https://github.com/simplereach/helenus/commit/aaca8447fcba7e07c44891a904a559fdf375a94d#diff-14 is included and corrected (`&&` instead of `||`)
- 57580be9962b2aed8df51317605b92c38bf9438c is included
- 5d38e53193e4b9199ba5eea1d83c578d09e021b5 is included
- 5c531e5ee1ea2db3acccbe32c68e1ec0e7183157 is included
- 1093bb468c7a4664ce74eded6e09f4bc88875648 is included
- 02869ce76cce13d68c3c40f1c9547d389f156f00 is included

Did I miss out anything?

devdazed commented:
there is another copy paste error here: https://github.com/simplereach/helenus/blob/aaca8447fcba7e07c44891a904a559fdf375a94d/lib/cassandra/1.0/cassandra_types.js#L1287

it looks like that is all the changes.  let me do some quick testing to see if it can all be pulled in.

ctavan commented:
Yeah, I got that one already: https://github.com/ctavan/helenus/compare/cql3-support#L4L1287

ctavan commented:
There's also a patch for thrift on https://issues.apache.org/jira/browse/THRIFT-1277 that should make parts of our patch redundant. I also think checking for `=== undefined` in our patch is not necessary.

ctavan committed:
Clean after compiling

devdazed commented:
I agree that the check for `!== undefined` is redundant as the classes all specifically set to `null` in their constructors.

ctavan committed:
Always check for `!== null` in cassandra_types

- Also set required start and finish options for slice predicates in the
  driver

ctavan commented:
I've changed all the checks to `!== null` which would also be fixed by the patch from [THRIFT-1277](https://issues.apache.org/jira/browse/THRIFT-1277).

That, however, had the consequence, that we must explicitly set some default value for the required `start` and `finish` parameter of the `SliceRange` structs. I think however that it's nicer to set these defaults in the node driver explicitly since that's what cassandra expects.

ctavan committed:
Re-generate stuff with thrift 0.7.0

ctavan committed:
Add tests for simple CQL3 support.

- Deserializers must now also be able to handle `null` as values since
  we can get column ghosts after deletion since CQL enforces the schema that
  has been defined.

ctavan committed:
Set CQL version explicitly and always

Cassandra keeps the CQL version after the connection is closed, so we set
the CQL version explicitly. That means however that you cannot use different
CQL versions with the same server.

ctavan commented:
I've translated the existing CQL tests to CQL3.

Deleting rows has now become even more quirky since you happen to see "column name ghosts" and not only "row key ghosts". But that's probably by design since CQL3 doesn't really distinguish between row keys and column names any more.

There seem to be some problems with composite columns (which are now natively supported in CQL) which I'll have to figure out and fix...

ctavan commented on lib/marshal/index.js:
```diff
@@ -178,7 +178,9 @@ function getDeserializer(type){
-    return TYPES[type].de;
+    return function(val) {
+      return val !== null ? TYPES[type].de(val) : null;
+    }
```
If we delete a row, the row key survives until the next compaction takes place. Since CQL3 enforces schemas, also all pre-defined column-keys will remain in that row as "column ghosts". They will however have `null` values regardless of their data type.

devdazed replied:
How will this change affect the Thrift interface part of it?

ctavan replied:
I think it shouldn't affect the Thrift interface since `val` will never be `null` when using the Thrift interface? Or am I getting something wrong? If you can find a case where this change breaks stuff, I'm happy to look into it :)

ctavan committed:
Don't set explicit default CQL version in the driver and split tests

- Rename CQL 2 tests accordingly
- Set the CQL version in the test files rather than in different configs

ctavan committed:
Remove trailing whitespaces

ctavan committed:
Rename all current CQL3 tests and reduce code duplication

- All currently included tests are for "static" column families

devdazed commented:
Does this include the fix from #17?

ctavan commented:
This branch includes all my other fixes. But I think we should wait a bit before we merge it since there are still some issues with composite columns that I want to fix here.

I'll rebase this branch against the master and include all other patches.

devdazed commented:
ok, let me know which branch I need to pull from to test all of the changes.

ctavan committed:
Add tests for "Standard "dynamic" CF"

- See https://issues.apache.org/jira/secure/attachment/12511286/create_cf_syntaxes.txt

ctavan committed:
Add convenience wrapper for performing tests on simple CQL queries.

ctavan committed:
Add support for CQL "dense composite" column families

- See
  https://issues.apache.org/jira/secure/attachment/12511286/create_cf_syntaxes.txt
  for an explanation of "dense composite"

ctavan committed:
Add tests for CQL "sparse composite" column families

- See
  https://issues.apache.org/jira/secure/attachment/12511286/create_cf_syntaxes.txt
  for an explanation of "sparse composite"

ctavan committed:
Add myself to the list of contributors ;)

ctavan committed:
Skip CQL3 tests if server doesnt support `set_cql_version()` call.

ctavan commented:
OK, I've rebased this one again. I'm done with it so far and I'd appreciate any feedback!

This branch now features support for CQL 3 queries as described and discussed in https://issues.apache.org/jira/browse/CASSANDRA-3761 . There are now 4 different types of column families (see https://issues.apache.org/jira/secure/attachment/12511286/create_cf_syntaxes.txt) which are now all supported by the driver.

The only CQL3 feature that is missing so far are prepared queries, but I think that's a task for a different pull request.

All new features can be tested against [cassandra-1.1.0~beta2](http://www.apache.org/dist/cassandra/debian/pool/main/c/cassandra/)

The tests also pass against cassandra 1.0.8 and 1.0.9.

ctavan commented:
Rebased again.
